proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,634 @@
+proposal aims to support most of the same things, but I will point out a
+few differences:
+
+- Export specifiers
+
+  - The export specifiers in [ghc283]_ are quite complex, and there are questions about how they should
+    behave for nested modules etc. The export specifiers in this
+    proposal are *extremely* simple: they just allow exporting a
+    module name. All additional structure must be added by structuring
+    the exported module.
+
+- Module aliases
+
+  - Module aliases barely make sense in the world of [ghc283]_
+    since there are no such things as modules.
+
+Conceptually, [ghc283]_ is a bit 
+like an “unrolled” version of this proposal, where we try and always
+work with the fully explicated list of qualified names. The problems
+above arise because we sometimes *want* to talk about modules
+themselves, but there are also problems with working with the full set
+of qualified names:
+
+- Recursive modules could cause non-termination.
+
+  - If ``A`` exports itself, then arguably it should export an
+    infinite set of qualified names: ``A``, ``A.A``, ``A.A.A``\ …
+  - [ghc283]_ avoids this by simply banning recursive local modules, but we don't
+    have to do this.
+  - See `Termination`_ below for why this isn’t a problem for this
+    proposal.
+
+Ultimately, I think this proposal is just *simpler*, in that it makes
+modules more like everything else in the module system, which then
+allows us to solve our problems with the namespacing tools we already
+have for other bindings.
+
+There is also an earlier proposal “Structured module exports/imports” [ghc205]_
+I won't respond to it in detail, but it takes the same approach of continuing to
+focus on qualified names rather than giving modules an identity.
+
+Unresolved questions
+--------------------
+
+What should happen if multiple modules are defined with the same name?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+[ghc283]_ effectively proposes that multiple modules with the same name should be
+*merged*. I would like to say that multiple modules with the same name
+should be *ambiguous*, just like all other names.
+
+I think the argument for merging is:
+
+- We have some existing cases where this happens (``import as``), which we need to support.
+- It might be convenient to “mix in” additional names into an existing module.
+
+The argument for ambiguity is:
+
+- Ambiguity is consistent with the way names work for everything else, so makes the system work.
+- It is not very inconvenient in practice
+
+  - Anecdotal evidence: I wrote the Semmle QL module system this way, and I never had anybody ask
+    for module merging.
+  - It is easily to “manually” merge modules if you have local module definitions.
+
+If we thought ambiguity was more consistent with "modules as entities", we could add a language
+extension `AmbiguousAs` which makes such `import as` statements produce ambiguity errors instead.
```

Comment:
I think we we all agree with this proposal as our road map, This and something like #220 would be good preparatory steps.

Reply:
Do we all agree that this is bad? In my experience people use this quite heavily, particularly for re-exporting, e.g.
```
module A (module Export) where
import B as Export
import C as Export
...
```

So I don't think removing it would be uncontroversial.

Reply:
sorry "if we" was what I meant to write, not "we we". Even if `AmbiguousAs` isn't deprecated, it would make clear what is sugar and what isn't.